### PR TITLE
clean up Basic constructors and diff code

### PR DESCRIPTION
**Pull request by spencerlyon2**

### CODE REVIEW

File: src/types.jl [hunk 1]
```diff
@@ -23,31 +23,37 @@ end
...
-function Basic(s::String)
+function Base.convert(::Type{Basic}, s::String)
     a = Basic()
     ccall((:basic_parse, :libsymengine), Void, (Ptr{Basic}, Ptr{Int8}), &a, s)
     return a
 end
 
-Basic(ex::Expr) = Basic(string(ex))
+Base.convert(::Type{Basic}, ex::Union{Symbol,Expr}) = Basic(string(ex))
+function Base.convert(::Type{Basic}, x::Real)
```
Comment (isuruf) [inline src/types.jl]: Can you replace `x::Real` with `x::Float64`? `Real` types with more precision should be converted into a `RealMPFR`.

Comment (spencerlyon2) [inline src/types.jl]: Good catch. will do

File: src/calculus.jl [hunk 2]
```diff
@@ -26,13 +27,15 @@ function diff(b1::SymbolicType, b2::SymbolicType, b3::SymbolicType)
...
+    diff(b1, (b2,b3,b4,b5...))
+
 ## mixed partials
 diff(ex::SymbolicType, bs::Tuple) = reduce((ex, x) -> diff(ex, x), ex, bs)
-diff(ex::SymbolicType, bs::Tuple,ns::Tuple) = reduce((ex, x) -> diff(ex, x[1],x[2]), ex, zip(bs,ns))
-
+diff(ex::SymbolicType, bs::Tuple, ns::Tuple) =
+    reduce((ex, x) -> diff(ex, x[1],x[2]), ex, zip(bs, ns))
 
+diff(b1::SymbolicType, x::Union{String,Symbol}) = diff(b1, Basic(x))
```
Comment (isuruf) [inline src/calculus.jl]: Since `String` is convertible to `Basic`, is this needed?

Comment (spencerlyon2) [inline src/calculus.jl]: I believe so because none of our `diff(b, x)` methods call `convert(Basic, x)`

**Comment by isuruf** [comment]
Can you change `Base.convert` to `convert`?  Otherwise +1 to merge

**Comment by spencerlyon2** [comment]
done!

**Pull request merged**